contracts/contracts/OpenAiChatGpt.sol

```diff
@@ -0,0 +1,213 @@
+        logitBias : "", // empty str for null
+        maxTokens : 1000, // 0 for null
+        presencePenalty : 21, // > 20 for null
+        responseFormat : "{\"type\":\"text\"}",
+        seed : 0, // null
+        stop : "", // null
+        temperature : 10, // Example temperature (scaled up, 10 means 1.0)
+        tools : "[{\"type\":\"function\",\"function\":{\"name\":\"web_search\",\"description\":\"Search the internet\",\"parameters\":{\"type\":\"object\",\"properties\":{\"query\":{\"type\":\"string\",\"description\":\"Search query\"}},\"required\":[\"query\"]}}}]",
+        toolChoice : "none",
+        user : "" // null
+        });
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Caller is not owner");
+        _;
+    }
+
+    modifier onlyOracle() {
+        require(msg.sender == oracleAddress, "Caller is not oracle");
+        _;
+    }
+
+    function setOracleAddress(address newOracleAddress) public onlyOwner {
+        oracleAddress = newOracleAddress;
+        emit OracleAddressUpdated(newOracleAddress);
+    }
+
+    function startChat(string memory message) public returns (uint i) {
+        ChatRun storage run = chatRuns[chatRunsCount];
+
+        run.owner = msg.sender;
+        Message memory newMessage;
+        newMessage.content = message;
+        newMessage.role = "user";
+        run.messages.push(newMessage);
+        run.messagesCount = 1;
+
+        uint currentId = chatRunsCount;
+        chatRunsCount = chatRunsCount + 1;
+
+        IOracle(oracleAddress).createOpenAiLlmCall(currentId, config);
+        emit ChatCreated(msg.sender, currentId);
+
+        return currentId;
+    }
+
+    function onOracleOpenAiLlmResponse(
+        uint runId,
+        IOracle.OpenAiResponse memory response,
+        string memory errorMessage
+    ) public onlyOracle {
+        ChatRun storage run = chatRuns[runId];
+        require(
+            keccak256(abi.encodePacked(run.messages[run.messagesCount - 1].role)) == keccak256(abi.encodePacked("user")),
+            "No message to respond to"
+        );
+
+        if (compareStrings(response.content, "")) {
+            IOracle(oracleAddress).createFunctionCall(runId, response.functionName, response.functionArguments);
+        } else {
+            Message memory newMessage;
+            newMessage.role = "assistant";
+            newMessage.content = response.content;
+            run.messages.push(newMessage);
+            run.messagesCount++;
+        }
+
+    }
+
+    function onOracleFunctionResponse(
+        uint runId,
+        string memory response,
+        string memory errorMessage
+    ) public onlyOracle {
+        ChatRun storage run = chatRuns[runId];
+        require(
+            compareStrings(run.messages[run.messagesCount - 1].role, "user"),
+            "No function to respond to"
+        );
+        if (compareStrings(errorMessage, "")) {
+            Message memory newMessage;
```

Comment:
Adds function response as a human message in this contract, and creates a new LLM call, where LLM has the function response in the context, ideally here config that is used would not use tools I guess